lib/Dialect/StaticLogic/Ops.cpp

```diff
@@ -19,9 +20,82 @@ using namespace circt::staticlogic;
 
 StaticLogicDialect::StaticLogicDialect(MLIRContext *context)
     : Dialect(getDialectNamespace(), context,
-    ::mlir::TypeID::get<StaticLogicDialect>()) {
+              ::mlir::TypeID::get<StaticLogicDialect>()) {
   addOperations<
 #define GET_OP_LIST
 #include "circt/Dialect/StaticLogic/StaticLogic.cpp.inc"
       >();
 }
+
+//===----------------------------------------------------------------------===//
+// PipelineOp
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// This canonicalizer will ensure all values passed between pipeline stages are
+/// going through the register operation.
+struct CanonicalizePipeline : public OpRewritePattern<PipelineOp> {
```

Comment:
Do you think this canonicalizer is really necessary?  I was thinking that the RegisterOp would purposefully not pass data, just to keep things simpler.  Is there a reason why you think the data has to be passed through the RegisterOp?

Reply:
I think this will make things easier when doing lowering. However, I agree with you that during the optimization / transformation, it's not necessary to ensure this. Not sure whether put this into the canonicalizer is reasonable.